### PR TITLE
Edge analysis

### PR DESCRIPTION
**Pull request by mjohnson541**
This PR ports RMG's species selection algorithms to RMS enabling us to do species selection on any system RMS can simulate.

### CODE REVIEW

File: src/EdgeAnalysis.jl [hunk 1]
```diff
@@ -22,4 +22,49 @@ end
...
+Generate appropriate core Simulation/SystemSimulation object
+"""
+function tosim(sol,domains::Tuple,inters,p)
+   return SystemSimulation(sol,domains,inters,p) 
+end
+
+"""
+Generate appropriate edge Simulation/SystemSimulation object
+"""
+function getsim(inte,react,coreedgedomain,domain,inters,p,coretoedgespcmap)
```
Comment (hwpang) [inline src/EdgeAnalysis.jl:44]: Input `domain` isn't used in this function? What is the difference between `coreedgedomain` and `domain`?

File: src/EdgeAnalysis.jl [hunk 2]
```diff
@@ -22,4 +22,49 @@ end
...
+    ycoreedge = getycoreedge(inte.u,coretoedgespcmap,coreedgedomain.indexes[end])
+    sol = build_solution(react.ode,inte.alg,[0.0,inte.t],[ycoreedge,ycoreedge])
+    sim = Simulation(sol,coreedgedomain,inters,p)
+    return sim
+end
+
+"""
+Generate appropriate edge Simulation/SystemSimulation object
+"""
+function getsim(inte,react,coreedgedomains::Tuple,domains::Tuple,inters,p,coretoedgespcmap)
```
Comment (hwpang) [inline src/EdgeAnalysis.jl:54]: Input `domains` isn't used in this function?

File: src/EdgeAnalysis.jl [hunk 3]
```diff
@@ -164,4 +164,85 @@ Calculate key flux and concentration related quantities for edge analysis
...
+    
+    return corespcsinds,corerxninds,edgespcsinds,edgerxninds,reactantindices,productindices,coretoedgespcmap,coretoedgerxnmap
+end
+
+"""
+Precalculate important indices and maps for use in edge analysis
+"""
+function getkeyselectioninds(coreedgedomain::AbstractDomain,coreedgeinters,domain,inters)
+    corespcsinds = 1:length(domain.phase.species)
+    edgespcsinds = length(domain.phase.species)+1:length(coreedgedomain.phase.species)
```
Comment (hwpang) [inline src/EdgeAnalysis.jl:237]: It looks like `coreedgedomain` actually only holds the edge species/reactions here, while in the testing part the coreedgedomain contains both of the core and edge species/reactions?

File: src/EdgeAnalysis.jl [hunk 4]
```diff
@@ -462,3 +462,206 @@ function updatefilterthresholds!(sim,corespcsinds,corespeciesconcentrations,char
...
+    breakflag = false
+    numcorespc = length(corespcsinds)
+    numcorerxns = length(corerxninds)
+    invalidobjectsprintboolean = true
+    terminated = false
+    
+    (dydt,rts,frts,rrts,cs,corespeciesratse,charrate,edgespeciesrates,
+    edgereactionrates,corespeciesrateratios,edgespeciesrateratios,
+    corereactionrates,corespeciesconcentrations,corespeciesproductionrates,
+    corespeciesconsumptionrates) = processfluxes(sim::Simulation,corespcsinds,corerxninds,edgespcsinds,edgerxninds)
```
Comment (hwpang) [inline src/EdgeAnalysis.jl:489]: Is `::Simulation` necessary here?

**Review by hwpang — COMMENTED** [review]
Thanks for the nice PR! Can't wait to use it! I had some comments and questions. Overall looks good to me.

File: src/TestEdgeAnalysis.jl [hunk 5]
```diff
@@ -0,0 +1,65 @@
+using Test
+using DiffEqBase
+using Sundials
+
+@testset "Species Selection" begin
+    outcore = readinput("../src/testing/minimal.rms")
+    outedge = readinput("../src/testing/minimal_edge.rms")
+    corespcs = outcore["phase"]["Species"]
+    edgeishspcs = outedge["phase"]["Species"]
```
Comment (hwpang) [inline src/TestEdgeAnalysis.jl:9]: Why `edgeishspcs`? Is this to deal with cases that has species both exist in core and edge input files?

File: src/EdgeAnalysis.jl [hunk 6]
```diff
@@ -462,3 +462,206 @@ function updatefilterthresholds!(sim,corespcsinds,corespeciesconcentrations,char
 end
 
 export updatefilterthresholds!
+
+"""
+Determine species/reactions that should be added to the model core, react thresholding and 
+whether the simulation should be interrupted or terminated
+"""
+function identifyobjects!(sim,simcore,corespcsinds,corerxninds,edgespcsinds,
```
Comment (hwpang) [inline src/EdgeAnalysis.jl:470]: `simcore` isn't used in this function?

File: src/Reactor.jl [hunk 7]
```diff
@@ -275,6 +275,48 @@ end
...
+            @inbounds @fastmath fR = kfs[i]*cs[rarray[1,i]]*cs[rarray[2,i]]*cs[rarray[3,i]]
+        end
+        if @inbounds rarray[5,i] == 0
+            @inbounds @fastmath rR = krevs[i]*cs[rarray[4,i]]
+        elseif @inbounds rarray[6,i] == 0
+            @inbounds @fastmath rR = krevs[i]*cs[rarray[4,i]]*cs[rarray[5,i]]
+        else
+            @inbounds @fastmath rR = krevs[i]*cs[rarray[4,i]]*cs[rarray[5,i]]*cs[rarray[6,i]]
+        end
+        @fastmath frts[i] = fR*V
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply src/Reactor.jl:297]: Yep! Thanks!

File: Manifest.toml [hunk 8]
```diff
@@ -610,51 +609,51 @@ uuid = "d00139f3-1899-568f-a2f0-47f597d42d70"
...
 [[Mmap]]
 uuid = "a63ad114-7e13-5084-954f-fe012c677804"
 
 [[ModelingToolkit]]
 deps = ["ArrayInterface", "DataStructures", "DiffEqBase", "DiffEqJump", "DiffRules", "Distributed", "DocStringExtensions", "GeneralizedGenerated", "IfElse", "LabelledArrays", "Latexify", "Libdl", "LightGraphs", "LinearAlgebra", "MacroTools", "NaNMath", "RecursiveArrayTools", "Requires", "SafeTestsets", "SparseArrays", "SpecialFunctions", "StaticArrays", "SymbolicUtils", "TreeViews", "UnPack", "Unitful"]
-git-tree-sha1 = "a2777851fd21be36882ddafe5b5e2036cdc14636"
+git-tree-sha1 = "c5bbf9990d0be177c3edeab96eb7551ad07c9553"
 uuid = "961ee093-0014-501f-94e3-6117800e7a78"
-version = "3.20.1"
+version = "3.21.0"
```
[the remark this answers is not in the document]
Comment (mjohnson541) [reply Manifest.toml:650]: The manifest file is updated automatically so my guess is that it's related to compatibility among our dependencies.

File: src/EdgeAnalysis.jl (hunk 1 above)
Comment (mjohnson541) [reply src/EdgeAnalysis.jl:44]: Thanks! coreedgedomain is the domain containing the core and edge while domain is the one containing just the core.

File: src/EdgeAnalysis.jl (hunk 2 above)
Comment (mjohnson541) [reply src/EdgeAnalysis.jl:54]: Thanks!

File: src/EdgeAnalysis.jl (hunk 3 above)
Comment (mjohnson541) [reply src/EdgeAnalysis.jl:237]: It contains both edge and core, but while it's ordered core,edge it doesn't know when edge starts so we determine that from the core domain. Does that make sense?

File: src/EdgeAnalysis.jl (hunk 4 above)
Comment (mjohnson541) [reply src/EdgeAnalysis.jl:489]: Nope, in fact that's not good. Thanks!

File: src/EdgeAnalysis.jl (hunk 6 above)
Comment (mjohnson541) [reply src/EdgeAnalysis.jl:470]: Thanks!

File: src/TestEdgeAnalysis.jl (hunk 5 above)
Comment (mjohnson541) [reply src/TestEdgeAnalysis.jl:9]: Yep, it isn't just the edge species.

File: src/EdgeAnalysis.jl (hunk 3 above)
Comment (hwpang) [reply src/EdgeAnalysis.jl:237]: Ah ok yeah that makes sense! Thanks!

File: src/TestEdgeAnalysis.jl (hunk 5 above)
Comment (hwpang) [reply src/TestEdgeAnalysis.jl:9]: Thanks! This is very useful

**Comment by mjohnson541** [comment]
Okay I think I've fixed everything.

**Pull request merged**